cpp/oneapi/dal/test/engine/common.hpp

```diff
@@ -0,0 +1,208 @@
+#define TEMPLATE_LIST_TEST   TEMPLATE_LIST_TEST_CASE
+#define TEMPLATE_SIG_TEST    TEMPLATE_TEST_CASE_SIG
+#define TEST_M               TEST_CASE_METHOD
+#define TEMPLATE_TEST_M      TEMPLATE_TEST_CASE_METHOD
+#define TEMPLATE_LIST_TEST_M TEMPLATE_LIST_TEST_CASE_METHOD
+#define TEMPLATE_SIG_TEST_M  TEMPLATE_TEST_CASE_METHOD_SIG
+
+#ifdef ONEDAL_DATA_PARALLEL
+#define DECLARE_TEST_POLICY(policy_name) oneapi::dal::test::engine::device_test_policy policy_name
+#else
+#define DECLARE_TEST_POLICY(policy_name) oneapi::dal::test::engine::host_test_policy policy_name
+#endif
+
+namespace oneapi::dal::test::engine {
+
+class host_test_policy {};
+
+template <typename... Args>
+inline auto train(host_test_policy& policy, Args&&... args) {
+    return dal::train(std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+inline auto infer(host_test_policy& policy, Args&&... args) {
+    return dal::infer(std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+inline auto compute(host_test_policy& policy, Args&&... args) {
+    return dal::compute(std::forward<Args>(args)...);
+}
+
+#ifdef ONEDAL_DATA_PARALLEL
+class test_queue_provider {
+public:
+    static test_queue_provider& get_instance();
+
+    sycl::queue& get_global_queue() {
+        if (!queue_) {
+            throw internal_error{ "Test queue provider is not initialized" };
+        }
+        return *queue_;
+    }
+
+    void init(const sycl::queue& queue) {
+        queue_.reset(new sycl::queue{ queue });
+    }
+
+    void reset() {
+        queue_.reset();
+    }
+
+private:
+    test_queue_provider() = default;
+
+    std::unique_ptr<sycl::queue> queue_;
+};
+
+inline const sycl::queue& get_global_queue() {
+    return test_queue_provider::get_instance().get_global_queue();
+}
+
+class device_test_policy {
+public:
+    device_test_policy(const sycl::queue& queue) : queue_(queue) {}
+
+    device_test_policy() : queue_(get_global_queue()) {}
+
+    sycl::queue& get_queue() {
+        return queue_;
+    }
+
+private:
+    sycl::queue queue_;
+};
+
+template <typename... Args>
+inline auto train(device_test_policy& policy, Args&&... args) {
+    return dal::train(policy.get_queue(), std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+inline auto infer(device_test_policy& policy, Args&&... args) {
+    return dal::infer(policy.get_queue(), std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+inline auto compute(device_test_policy& policy, Args&&... args) {
+    return dal::compute(policy.get_queue(), std::forward<Args>(args)...);
+}
+#endif
+
+template <std::size_t index, typename TupleX, typename TupleY>
+struct combine_types_element {
+private:
+    static constexpr std::size_t count_y = std::tuple_size_v<TupleY>;
+    static constexpr std::size_t i = index / count_y;
+    static constexpr std::size_t j = index % count_y;
+
+public:
+    using type = std::tuple<std::tuple_element_t<i, TupleX>, std::tuple_element_t<j, TupleY>>;
```

Comment:
TupleX offset is calculated from TupleY. Is it necessary to check sizes?

Reply:
I can put static assert here